src/app/libp2p_helper/src/libp2p_helper/main.go

```diff
@@ -656,40 +658,55 @@ type incomingMsgUpcall struct {
 }
 
 func handleStreamReads(app *app, stream net.Stream, idx int) {
+	// create buffer stream for non-blocking read
+	r := bufio.NewReader(stream)
+	msg := app.P2p.MsgStats
+
+	var msgBytes []byte
 	go func() {
 		defer func() {
 			_ = stream.Close()
 		}()
-
-		buf := make([]byte, 4096)
 		for {
-			len, err := stream.Read(buf)
-
-			if len != 0 {
-				app.writeMsg(incomingMsgUpcall{
-					Upcall:    "incomingStreamMsg",
-					Data:      codaEncode(buf[:len]),
+			length, err := readLEB128ToUint64(r)
+			if err == io.EOF {
+				continue
+			} else if err != nil {
+				app.writeMsg(streamLostUpcall{
+					Upcall:    "streamLost",
 					StreamIdx: idx,
+					Reason:    fmt.Sprintf("read failure: %s", err.Error()),
 				})
+				return
+			}
+
+			if length == 0 {
+				continue
 			}
 
-			if err != nil && err != io.EOF {
+			msg.UpdateMetrics(length)
+
+			msgBytes = make([]byte, length)
```

Comment:
Seems bad to allocate an array with length provided by a peer -- is there a way we can have a bounded size here?

Reply:
Done in #7724